### PR TITLE
Developer: Network Debugging

### PR DESCRIPTION
**Pull request by bjtitus**
Adds [Pulse](https://github.com/kean/Pulse) network logging framework.

| | |
|-|-|
|  |  |

https://github.com/Automattic/pocket-casts-ios/assets/3250/39de84de-9cf2-4205-a162-141cb8ac2bc3

### Network debugging
* Enable the `networkDebugging` flag in Beta Features
* Restart the app
* Download an episode
* Navigate to Profile > Settings > Developer > Network Debugging
* Note that episode download, along with other network activity and logs, is shown in debugging menu

### Exported Logs
* Profile > Help & Feedback > Export Database
* Follow the instructions below to access export:
  * **Device**: Airdrop the file to your computer 
  * **Simulator**: Run this script to access the files from Files.app:
```bash
#!/bin/bash
open "$(xcrun simctl get_app_container booted com.apple.DocumentsApp groups | grep LocalStorage | awk -F'\t' '{print $2 "/File Provider Storage"}')"
```
* Verify that `network.pulse` file exists
* This file can be opened with the [Pulse Pro](https://pulselogger.com/) mac app

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/DataModel/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages in Xcode.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Server/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages in Xcode.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Utils/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages in Xcode.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by pocketcasts** [comment]
<details><summary>Pocket Casts Prototype Build📲 You can test the changes from this Pull Request in <b>Pocket Casts Prototype Build</b> by scanning the QR code below to install the corresponding build.</summary>
<table>
<tr>
  <td rowspan='6' width='260px'></td>
  <td><b>App Name</b></td><td>Pocket Casts Prototype Build Pocket Casts Prototype Build</td>
</tr>
<tr><td><b>Build Number</b></td><td>pr1813-e437fe9</td></tr>
<tr><td><b>Version</b></td><td>7.66</td></tr>
<tr><td><b>Bundle ID</b></td><td>au.com.shiftyjelly.podcasts.prototype</td></tr>
<tr><td><b>Commit</b></td><td>e437fe959c4c32062603a23693e5eb5315c285de</td></tr>
<tr><td><b>App Center Build</b></td><td><a href='https://install.appcenter.ms/orgs/automattic/apps/Pocket-Casts-Prototype-Builds/releases/13'>Pocket Casts - Prototype Builds #13</a></td></tr>
</table>
<em>Automatticians: You can use our internal self-serve MC tool to give yourself access to App Center if needed.</em>
</details>

File: Podfile [hunk 1]
```diff
@@ -21,6 +21,8 @@ end
 target 'podcasts' do
   platform :ios, app_ios_deployment_target.version
   common_pods
+  pod 'PulseCore', :git => 'https://github.com/kean/Pulse.git', :tag => '4.2.4', :configurations => ['Debug', 'Staging', 'Prototype']
```
Comment (dangermattic) [inline Podfile:24]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/f480240a65531a6607ddef00a3260a1d14d1a519/Podfile#L24"></span>Style/HashSyntax: Use the new Ruby 1.9 hash syntax.</td>
    </tr>
  </tbody>
</table>

Comment (dangermattic) [inline Podfile:24]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/f480240a65531a6607ddef00a3260a1d14d1a519/Podfile#L24"></span>Style/WordArray: Use <code>%w</code> or <code>%W</code> for an array of words.</td>
    </tr>
  </tbody>
</table>

File: Podfile [hunk 2]
```diff
@@ -21,6 +21,8 @@ end
...
+  pod 'PulseUI', :git => 'https://github.com/kean/Pulse.git', :tag => '4.2.4', :configurations => ['Debug', 'Staging', 'Prototype']
```
Comment (dangermattic) [inline Podfile:25]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/f480240a65531a6607ddef00a3260a1d14d1a519/Podfile#L25"></span>Style/HashSyntax: Use the new Ruby 1.9 hash syntax.</td>
    </tr>
  </tbody>
</table>

Comment (dangermattic) [inline Podfile:25]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/f480240a65531a6607ddef00a3260a1d14d1a519/Podfile#L25"></span>Style/WordArray: Use <code>%w</code> or <code>%W</code> for an array of words.</td>
    </tr>
  </tbody>
</table>

**Comment by bjtitus** [comment]
@leandroalonso I tried several different ways to reference the Pulse Swift Package only in Release builds but none of them seemed to work. I've just moved it to Cocoapods for now. I verified it wasn't included by checking for symbols using Hopper:

CleanShot 2024-06-13 at 17 30 47@2x

File: Podfile [hunk 3]
```diff
@@ -21,6 +21,8 @@ end
 target 'podcasts' do
   platform :ios, app_ios_deployment_target.version
   common_pods
+  pod 'PulseCore', :git => 'https://github.com/kean/Pulse.git', :tag => '4.2.4', :configurations => %w(Debug Staging Prototype)
```
Comment (dangermattic) [inline Podfile:24]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/456e5b941c35f04ba523a63c2f5644e9734320a8/Podfile#L24"></span>Style/HashSyntax: Use the new Ruby 1.9 hash syntax.</td>
    </tr>
  </tbody>
</table>

File: Podfile [hunk 4]
```diff
@@ -21,6 +21,8 @@ end
...
+  pod 'PulseUI', :git => 'https://github.com/kean/Pulse.git', :tag => '4.2.4', :configurations => %w(Debug Staging Prototype)
```
Comment (dangermattic) [inline Podfile:25]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/456e5b941c35f04ba523a63c2f5644e9734320a8/Podfile#L25"></span>Style/HashSyntax: Use the new Ruby 1.9 hash syntax.</td>
    </tr>
  </tbody>
</table>

Comment (dangermattic) [inline Podfile:25]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/456e5b941c35f04ba523a63c2f5644e9734320a8/Podfile#L25"></span>Style/PercentLiteralDelimiters: <code>%w</code>-literals should be delimited by <code>[</code> and <code>]</code>.</td>
    </tr>
  </tbody>
</table>

**Comment by leandroalonso** [comment]
@bjtitus Nice! The test target is failing though. 🔴

**Comment by bjtitus** [comment]
Good to go now @leandroalonso and @SergioEstevao!

**Review by SergioEstevao — COMMENTED** [review]
Great addition to our debug tools!
It could be very helpful to find issues when streaming files and see request results on the context of playback skipping.

In general the tool is working correctly and I just found a crash but I think it's on Pulse codebase.
Steps to reproduce:
- Go to Network Debug
- Tap on Filters and select group by host
- On a Group tap on See All, 
- See it crash

**Review by SergioEstevao — APPROVED** [review]
Looking good, just left a comment regarding a crash that I think is part of the PulseUI library itself.

**Comment by bjtitus** [comment]
Thanks @SergioEstevao! I reproduced the crash and filed an issue (https://github.com/kean/Pulse/issues/263) on Pulse.

**Pull request merged**